[C++ Frontend] Fix Windows build and test in CI

goldsborough opened the pull request:
This PR adds Windows support for the C++ frontend. A lot of declarations were missing `TORCH_API` macros, and lots of code just did not compile on MSVC.

@ebetica @ezyang @orionr

peterjc123 commented on .jenkins/pytorch/win-build.sh:
```diff
@@ -137,11 +144,11 @@ if not "%USE_CUDA%"=="0" (
   if "%REBUILD%"=="" set NO_CUDA=0
 
-  python setup.py install && sccache --show-stats && (
+  python setup.py install && call ci_scripts/build_libtorch.bat && sccache --show-stats && (
```
Remove ci_scripts since we are already under that dir.

peterjc123 commented on tools/build_libtorch.py:
```diff
@@ -18,7 +19,10 @@
 
     tools_path = os.path.dirname(os.path.abspath(__file__))
-    build_pytorch_libs = os.path.join(tools_path, 'build_pytorch_libs.sh')
+    if options.windows:
```
Is there a reason that we don't just use `IS_WINDOWS` here?

goldsborough committed:
[C++ Frontend] Support Windows for the C++ frontend

[C++ Frontend] Fix Windows builid and test in CI

Need some debuginfo

Zeming's fixes for Windoze

Fix windows macro

checkpoint

Fix any

More fixes

goldsborough commented on c10/util/Optional.h:
```diff
@@ -361,7 +361,7 @@ class optional : private OptionalBase<T> {
 
   void clear() noexcept {
     if (initialized())
-      dataptr()->T::~T();
```
@ezyang

orionr commented:
Excellent. Let's wait until https://github.com/pytorch/pytorch/pull/13100 lands before bringing this in.

orionr commented:
cc @davidbrownellMS

orionr commented on tools/build_pytorch_libs.bat:
```diff
@@ -212,7 +212,7 @@ goto:eof
                   -DBUILD_TORCH="%BUILD_TORCH%" ^
                   -DNVTOOLEXT_HOME="%NVTOOLEXT_HOME%" ^
-                  -DNO_API=ON ^
+                  -DNO_API=OFF ^
```
You can just remove this line instead I think.

goldsborough replied:
ok

orionr commented on torch/CMakeLists.txt:
```diff
@@ -344,6 +343,10 @@ endif()
 
 if(USE_CUDA)
   if(MSVC)
+    set(NVTOOLEXT_HOME "C:/Program Files/NVIDIA Corporation/NvToolsExt")
```
This hardcoded path makes me nervous... :-/ Any way around it. @davidbrownellMS and @peterjc123 any thoughts?

peterjc123 replied:
I think it should be okay. This path is rather constant. If we want to be safe, then we'll need to do reg query.

goldsborough committed:
More missing TORCH_API macros

orionr commented on torch/csrc/api/src/nn/module.cpp:
```diff
@@ -62,6 +62,14 @@ const std::string& Module::name() const noexcept {
   if (!name_.has_value()) {
     name_ = c10::demangle(typeid(*this).name());
+#if defined(_WIN32)
+    // Windows adds "struct" or "class" as a prefix.
```
Ouch. any way to get around this special case? cc @davidbrownellMS and @peterjc123 . If not, so be it.

peterjc123 replied:
[That](https://www.boost.org/doc/libs/1_66_0/boost/dll/detail/demangling/msvc.hpp)'s how Boost deals with this. So I think it should be fine.

orionr reviewed (changes requested):
Great stuff. I think the hardcoded path is the one thing to talk about. Also, can we rebase this on https://github.com/pytorch/pytorch/pull/13100 once it lands? Should be in progress right now.

goldsborough commented on aten/src/ATen/templates/GeneratorDerived.h:
```diff
@@ -11,16 +11,16 @@ namespace at {
 struct ${name}Generator : public Generator {
   ${name}Generator(Context * context);
-  virtual ~${name}Generator();
+  CAFFE2_API virtual ~${name}Generator();
```
That complained about a missing `vftable` unfortunately, that's why I put it at function level

orionr replied:
@peterjc123 any thoughts?

peterjc123 replied:
Are there any function implementations in this header?

goldsborough committed:
Fix WindowsTorchApiMacro include

orionr commented:
Referenced PR just landed - would you mind rebasing? Much appreciated.

Commit:
More TORCH_API

Commit:
Add torch_api to scatter/gather

Commit:
Make NVTools include public

Commit:
Send test_api.exe to win-test.sh via 7z?

Commit:
Fix win-test.sh

Commit:
#ifdef out Tempfile test on Windows

ebetica reviewed (approved):
:rocket:

peterjc123 commented:
Lookings like it's breaking the master branch in the CI of Azure Pipelines.
```
2018-11-14T04:38:58.2954015Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: __cdecl onnx_torch::AttributeProto::AttributeProto(class onnx_torch::AttributeProto &&)" (??0AttributeProto@onnx_torch@@QEAA@$$QEAV01@@Z) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2954993Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: __cdecl onnx_torch::AttributeProto::AttributeProto(class onnx_torch::AttributeProto const &)" (??0AttributeProto@onnx_torch@@QEAA@AEBV01@@Z) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2955350Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: __cdecl onnx_torch::AttributeProto::AttributeProto(void)" (??0AttributeProto@onnx_torch@@QEAA@XZ) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2955616Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: __cdecl onnx_torch::GraphProto::GraphProto(void)" (??0GraphProto@onnx_torch@@QEAA@XZ) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2955894Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: __cdecl onnx_torch::ModelProto::ModelProto(class onnx_torch::ModelProto &&)" (??0ModelProto@onnx_torch@@QEAA@$$QEAV01@@Z) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2956278Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: __cdecl onnx_torch::ModelProto::ModelProto(class onnx_torch::ModelProto const &)" (??0ModelProto@onnx_torch@@QEAA@AEBV01@@Z) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2956646Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: __cdecl onnx_torch::ModelProto::ModelProto(void)" (??0ModelProto@onnx_torch@@QEAA@XZ) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2956939Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: __cdecl onnx_torch::NodeProto::NodeProto(class onnx_torch::NodeProto const &)" (??0NodeProto@onnx_torch@@QEAA@AEBV01@@Z) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2957211Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: __cdecl onnx_torch::NodeProto::NodeProto(void)" (??0NodeProto@onnx_torch@@QEAA@XZ) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2962533Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: __cdecl onnx_torch::OperatorSetIdProto::OperatorSetIdProto(void)" (??0OperatorSetIdProto@onnx_torch@@QEAA@XZ) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2962844Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: __cdecl onnx_torch::StringStringEntryProto::StringStringEntryProto(void)" (??0StringStringEntryProto@onnx_torch@@QEAA@XZ) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2963399Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: __cdecl onnx_torch::TensorProto::TensorProto(void)" (??0TensorProto@onnx_torch@@QEAA@XZ) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2963687Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: __cdecl onnx_torch::TensorProto_Segment::TensorProto_Segment(void)" (??0TensorProto_Segment@onnx_torch@@QEAA@XZ) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2964077Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: __cdecl onnx_torch::TensorShapeProto::TensorShapeProto(class onnx_torch::TensorShapeProto const &)" (??0TensorShapeProto@onnx_torch@@QEAA@AEBV01@@Z) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2964368Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: __cdecl onnx_torch::TensorShapeProto::TensorShapeProto(void)" (??0TensorShapeProto@onnx_torch@@QEAA@XZ) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2964704Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: __cdecl onnx_torch::TensorShapeProto_Dimension::TensorShapeProto_Dimension(class onnx_torch::TensorShapeProto_Dimension &&)" (??0TensorShapeProto_Dimension@onnx_torch@@QEAA@$$QEAV01@@Z) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2965636Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: __cdecl onnx_torch::TensorShapeProto_Dimension::TensorShapeProto_Dimension(class onnx_torch::TensorShapeProto_Dimension const &)" (??0TensorShapeProto_Dimension@onnx_torch@@QEAA@AEBV01@@Z) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2965995Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: __cdecl onnx_torch::TensorShapeProto_Dimension::TensorShapeProto_Dimension(void)" (??0TensorShapeProto_Dimension@onnx_torch@@QEAA@XZ) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2966437Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: __cdecl onnx_torch::TypeProto::TypeProto(class onnx_torch::TypeProto &&)" (??0TypeProto@onnx_torch@@QEAA@$$QEAV01@@Z) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2966733Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: __cdecl onnx_torch::TypeProto::TypeProto(class onnx_torch::TypeProto const &)" (??0TypeProto@onnx_torch@@QEAA@AEBV01@@Z) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2967002Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: __cdecl onnx_torch::TypeProto::TypeProto(void)" (??0TypeProto@onnx_torch@@QEAA@XZ) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2967777Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: __cdecl onnx_torch::TypeProto_Tensor::TypeProto_Tensor(void)" (??0TypeProto_Tensor@onnx_torch@@QEAA@XZ) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2968144Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: __cdecl onnx_torch::ValueInfoProto::ValueInfoProto(class onnx_torch::ValueInfoProto const &)" (??0ValueInfoProto@onnx_torch@@QEAA@AEBV01@@Z) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2968439Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: __cdecl onnx_torch::ValueInfoProto::ValueInfoProto(void)" (??0ValueInfoProto@onnx_torch@@QEAA@XZ) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2968682Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: virtual __cdecl onnx_torch::AttributeProto::~AttributeProto(void)" (??1AttributeProto@onnx_torch@@UEAA@XZ) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2969074Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: virtual __cdecl onnx_torch::GraphProto::~GraphProto(void)" (??1GraphProto@onnx_torch@@UEAA@XZ) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2969382Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: virtual __cdecl onnx_torch::ModelProto::~ModelProto(void)" (??1ModelProto@onnx_torch@@UEAA@XZ) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2969814Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: virtual __cdecl onnx_torch::NodeProto::~NodeProto(void)" (??1NodeProto@onnx_torch@@UEAA@XZ) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2970073Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: virtual __cdecl onnx_torch::TensorProto_Segment::~TensorProto_Segment(void)" (??1TensorProto_Segment@onnx_torch@@UEAA@XZ) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2972950Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: virtual __cdecl onnx_torch::TensorShapeProto::~TensorShapeProto(void)" (??1TensorShapeProto@onnx_torch@@UEAA@XZ) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2973229Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: virtual __cdecl onnx_torch::TensorShapeProto_Dimension::~TensorShapeProto_Dimension(void)" (??1TensorShapeProto_Dimension@onnx_torch@@UEAA@XZ) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2973639Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: virtual __cdecl onnx_torch::TypeProto::~TypeProto(void)" (??1TypeProto@onnx_torch@@UEAA@XZ) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2973963Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: virtual __cdecl onnx_torch::ValueInfoProto::~ValueInfoProto(void)" (??1ValueInfoProto@onnx_torch@@UEAA@XZ) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2974272Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: virtual void __cdecl onnx_torch::AttributeProto::Clear(void)" (?Clear@AttributeProto@onnx_torch@@UEAAXXZ) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2974599Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: virtual void __cdecl onnx_torch::OperatorSetIdProto::Clear(void)" (?Clear@OperatorSetIdProto@onnx_torch@@UEAAXXZ) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2974915Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: virtual void __cdecl onnx_torch::TensorShapeProto_Dimension::Clear(void)" (?Clear@TensorShapeProto_Dimension@onnx_torch@@UEAAXXZ) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2975851Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: virtual void __cdecl onnx_torch::TypeProto::Clear(void)" (?Clear@TypeProto@onnx_torch@@UEAAXXZ) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2976319Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: void __cdecl onnx_torch::AttributeProto::CopyFrom(class onnx_torch::AttributeProto const &)" (?CopyFrom@AttributeProto@onnx_torch@@QEAAXAEBV12@@Z) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2976635Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: void __cdecl onnx_torch::GraphProto::CopyFrom(class onnx_torch::GraphProto const &)" (?CopyFrom@GraphProto@onnx_torch@@QEAAXAEBV12@@Z) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2976934Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: void __cdecl onnx_torch::NodeProto::CopyFrom(class onnx_torch::NodeProto const &)" (?CopyFrom@NodeProto@onnx_torch@@QEAAXAEBV12@@Z) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2977205Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: void __cdecl onnx_torch::TensorProto::CopyFrom(class onnx_torch::TensorProto const &)" (?CopyFrom@TensorProto@onnx_torch@@QEAAXAEBV12@@Z) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2977523Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: void __cdecl onnx_torch::TensorProto_Segment::CopyFrom(class onnx_torch::TensorProto_Segment const &)" (?CopyFrom@TensorProto_Segment@onnx_torch@@QEAAXAEBV12@@Z) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2977927Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: void __cdecl onnx_torch::TensorShapeProto::CopyFrom(class onnx_torch::TensorShapeProto const &)" (?CopyFrom@TensorShapeProto@onnx_torch@@QEAAXAEBV12@@Z) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2978934Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: void __cdecl onnx_torch::TensorShapeProto_Dimension::CopyFrom(class onnx_torch::TensorShapeProto_Dimension const &)" (?CopyFrom@TensorShapeProto_Dimension@onnx_torch@@QEAAXAEBV12@@Z) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2979559Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: void __cdecl onnx_torch::TypeProto::CopyFrom(class onnx_torch::TypeProto const &)" (?CopyFrom@TypeProto@onnx_torch@@QEAAXAEBV12@@Z) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2979951Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: void __cdecl onnx_torch::TypeProto_Tensor::CopyFrom(class onnx_torch::TypeProto_Tensor const &)" (?CopyFrom@TypeProto_Tensor@onnx_torch@@QEAAXAEBV12@@Z) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2980335Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "private: void __cdecl onnx_torch::TensorShapeProto_Dimension::InternalSwap(class onnx_torch::TensorShapeProto_Dimension *)" (?InternalSwap@TensorShapeProto_Dimension@onnx_torch@@AEAAXPEAV12@@Z) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2987560Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "public: void __cdecl onnx_torch::TensorShapeProto_Dimension::MergeFrom(class onnx_torch::TensorShapeProto_Dimension const &)" (?MergeFrom@TensorShapeProto_Dimension@onnx_torch@@QEAAXAEBV12@@Z) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2987923Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "bool __cdecl onnx_torch::TensorProto_DataType_IsValid(int)" (?TensorProto_DataType_IsValid@onnx_torch@@YA_NH@Z) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2988192Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "private: void __cdecl onnx_torch::TensorShapeProto_Dimension::clear_value(void)" (?clear_value@TensorShapeProto_Dimension@onnx_torch@@AEAAXXZ) already defined in torch.lib(torch.dll)
2018-11-14T04:38:58.2988491Z onnx_proto.lib(onnx_onnx_torch.pb.cc.obj) : error LNK2005: "private: void __cdecl onnx_torch::TypeProto::clear_value(void)" (?clear_value@TypeProto@onnx_torch@@AEAAXXZ) already defined in torch.lib(torch.dll)
```

orionr commented:
@goldsborough can you work with @peterjc123 on this? Much appreciated.

orionr commented:
And FYI the above looks like the ONNX symbols are being declared in multiple places. I wonder if https://github.com/pytorch/pytorch/pull/13550 will help?

goldsborough commented:
@peterjc123 what is Azure Pipelines? Where can I see these logs?

peterjc123 commented:
@goldsborough It's the windows nightly CI. The log is available [here](https://dev.azure.com/peterjc123/pytorch_builder/_build/results?buildId=6&view=logs), just click on any one of the Windows CPU jobs and pick the Build stage.
